src/libproc/proc_pid.rs

```diff
@@ -637,3 +637,452 @@ impl From<uint32_t> for ProcFDType {
+///     }
+/// }
+/// ```
+///
+pub fn pidfdinfo<T: PIDFDInfo>(pid : i32, fd: int32_t) -> Result<T, String> {
+    let flavor = T::flavor() as i32;
+    let buffer_size = mem::size_of::<T>() as i32;
+    let mut pidinfo = T::default();
+    let buffer_ptr = &mut pidinfo as *mut _ as *mut c_void;
+    let ret: i32;
+
+    unsafe {
+        ret = proc_pidfdinfo(pid, fd, flavor, buffer_ptr, buffer_size);
+    };
+
+    if ret <= 0 {
+        Err(get_errno_with_message(ret))
+    } else {
+        Ok(pidinfo)
+    }
+}
+
+#[test]
+fn pidfdinfo_test() {
+    use std::process;
+    use std::net::TcpListener;
+    let pid = process::id() as i32;
+
+    let _listener = TcpListener::bind("127.0.0.1:65535");
+
+    match pidinfo::<BSDInfo>(pid, 0) {
+        Ok(info) => {
+            match listpidinfo::<ListFDs>(pid, info.pbi_nfiles as usize) {
+                Ok(fds) => {
+                    for fd in fds {
+                        match fd.proc_fdtype.into() {
+                            ProcFDType::Socket => {
+                                if let Ok(socket) = pidfdinfo::<SocketFDInfo>(pid, fd.proc_fd) {
+                                    match socket.psi.soi_kind.into() {
+                                        SocketInfoKind::Tcp => unsafe {
+                                            let info = socket.psi.soi_proto.pri_tcp;
+                                            assert_eq!(socket.psi.soi_protocol,
+                                            libc::IPPROTO_TCP);
+                                            assert_eq!(info.tcpsi_ini.insi_lport as u32, 65535);
+                                        }
+                                        _ => (),
+                                    }
+                                }
+                            },
+                            _ => (),
+                        }
+                    }
+                },
+                Err(err) => assert!(false, "Error retrieving process info: {}", err)
+            }
+        },
+        Err(err) => assert!(false, "Error retrieving process info: {}", err)
+    };
+}
+
+#[repr(C)]
+#[derive(Default)]
+pub struct SocketFDInfo {
+    pub pfi: ProcFileInfo,
+    pub psi: SocketInfo,
+}
+
+impl PIDFDInfo for SocketFDInfo {
+    fn flavor() -> PidFDInfoFlavor { PidFDInfoFlavor::SocketInfo }
+}
+
+#[repr(C)]
+#[derive(Default)]
+pub struct ProcFileInfo {
+    pub fi_openflags: uint32_t,
+    pub fi_status   : uint32_t,
+    pub fi_offset   : off_t,
+    pub fi_type     : int32_t,
+    pub rfu_1       : int32_t,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum SocketInfoKind {
+    Generic   = 0,
+    /// IPv4 and IPv6 Sockets
+    In        = 1,
+    /// TCP Sockets
+    Tcp       = 2,
+    /// Unix Domain Sockets
+    Un        = 3,
+    /// PF_NDRV Sockets
+    Ndrv      = 4,
+    /// Kernel Event Sockets
+    KernEvent = 5,
+    /// Kernel Control Sockets
+    KernCtl   = 6,
+    /// Unknown
+    Unknown,
+}
+
+impl From<c_int> for SocketInfoKind {
```

Comment:
It's a pity rust can't derive this type of stuff for us, to go from value -> enum, like you can in Java...

Reply:
Yes, I think so...